pandas/_libs/tslib.pyx

```diff
@@ -455,18 +455,18 @@ cpdef array_to_datetime(
         set out_tzoffset_vals = set()
         tzinfo tz_out = None
         bint found_tz = False, found_naive = False
-        cnp.broadcast mi
+        cnp.flatiter it = cnp.PyArray_IterNew(values)
 
     # specify error conditions
     assert is_raise or is_ignore or is_coerce
 
     result = np.empty((<object>values).shape, dtype="M8[ns]")
-    mi = cnp.PyArray_MultiIterNew2(result, values)
     iresult = result.view("i8").ravel()
 
     for i in range(n):
         # Analogous to `val = values[i]`
-        val = <object>(<PyObject**>cnp.PyArray_MultiIter_DATA(mi, 1))[0]
+        val = cnp.PyArray_GETITEM(values, cnp.PyArray_ITER_DATA(it))
+        cnp.PyArray_ITER_NEXT(it)
```

Comment:
Similar to comment above but worth closer inspection we should be using the iteration function https://numpy.org/doc/stable/reference/c-api/iterator.html#functions-for-iteration